src/tls.c

```diff
@@ -7007,86 +7008,94 @@ static int TLSX_KeyShare_GenEccKey(WOLFSSL *ssl, KeyShareEntry* kse)
         case WOLFSSL_ECC_SECP521R1:
             curveId = ECC_SECP521R1;
             keySize = 66;
-            dataSize = keySize * 2 + 1;
             break;
         #endif /* !NO_ECC_SECP */
-    #endif
-    #if defined(HAVE_X448) && ECC_MIN_KEY_SZ <= 448
-        case WOLFSSL_ECC_X448:
-            curveId = ECC_X448;
-            dataSize = keySize = 56;
-            break;
     #endif
         default:
             return BAD_FUNC_ARG;
     }
 
-    /* Allocate an ECC key to hold private key. */
-    keyPtr = (byte*)XMALLOC(sizeof(ecc_key), ssl->heap,
+    if (eccKey == NULL) {
+        kse->keyLen = keySize;
+        kse->pubKeyLen = keySize * 2 + 1;
+
+        /* Allocate an ECC key to hold private key. */
+        kse->key = (byte*)XMALLOC(sizeof(ecc_key), ssl->heap,
                                                       DYNAMIC_TYPE_PRIVATE_KEY);
-    if (keyPtr == NULL) {
-        WOLFSSL_MSG("EccTempKey Memory error");
-        return MEMORY_E;
-    }
-    eccKey = (ecc_key*)keyPtr;
+        if (kse->key == NULL) {
+            WOLFSSL_MSG("EccTempKey Memory error");
+            return MEMORY_E;
+        }
+        eccKey = (ecc_key*)kse->key;
 
-    /* Make an ECC key. */
-    ret = wc_ecc_init_ex(eccKey, ssl->heap, ssl->devId);
-    if (ret != 0)
-        goto end;
+        /* Make an ECC key */
+        ret = wc_ecc_init_ex(eccKey, ssl->heap, ssl->devId);
+        if (ret != 0) {
+            XFREE(kse->key, ssl->heap, DYNAMIC_TYPE_PRIVATE_KEY);
+            kse->key = NULL;
+            return ret;
+        }
 
-#ifdef WOLFSSL_STATIC_EPHEMERAL
-    if (ssl->staticKE.ecKey) {
-        DerBuffer* keyDer = ssl->staticKE.ecKey;
-        word32 idx = 0;
-        WOLFSSL_MSG("Using static ECDH key");
-        ret = wc_EccPrivateKeyDecode(keyDer->buffer, &idx, eccKey, keyDer->length);
-    }
-    else
-#endif
-    {
-        /* Generate ephemeral ECC key */
-        ret = wc_ecc_make_key_ex(ssl->rng, keySize, eccKey, curveId);
-    #ifdef WOLFSSL_ASYNC_CRYPT
-        /* TODO: Make this function non-blocking */
-        if (ret == WC_PENDING_E) {
-            ret = wc_AsyncWait(ret, &eccKey->asyncDev, WC_ASYNC_FLAG_NONE);
+        if (ret == 0) {
+        #ifdef WOLFSSL_STATIC_EPHEMERAL
+            if (ssl->staticKE.ecKey) {
+                DerBuffer* keyDer = ssl->staticKE.ecKey;
+                word32 idx = 0;
+                WOLFSSL_MSG("Using static ECDH key");
+                ret = wc_EccPrivateKeyDecode(keyDer->buffer, &idx, eccKey, keyDer->length);
+            }
+            else
+        #endif
+            {
+                /* set curve info for EccMakeKey "peer" info */
+                ret = wc_ecc_set_curve(eccKey, kse->keyLen, curveId);
+                if (ret == 0) {
+                    /* Generate ephemeral ECC key */
+                    ret = EccMakeKey(ssl, eccKey, eccKey);
```

Comment:
Re-entrant code needs to be isolated?

Reply:
Same. Not re entrant code. Called and it's the poll and higher level function that handles the async response code.